test/files/neg/t13006.scala

```diff
@@ -0,0 +1,26 @@
+//> using options -deprecation -Werror
+
+@deprecated case class C[T](x: T)
+case class D @deprecated() (x: Int)
+
+object A {
+  type X[T] = C[T] // warn
+  val X = C        // no warn
```

Comment:
if this doesn't warn, then merely aliasing "launders" the deprecation.

Reply:
That's true. But the fact that we issue a deprecation for the class during case-apply rewrite is also somewhat arbitrary.

```
scala> object o {
     |   @deprecated case class C(x: Int)
     |   @deprecated case class D private[o] (x: Int)
     | }
object o

scala> o.C(1)
         ^
       warning: class C in object o is deprecated
val res0: o.C = C(1)

scala> o.D(1)
val res1: o.D = D(1)
```

Reply:
Isn't that a bug? I assume the synthetic apply is not marked deprecated.

Or is arbitrary a synonym for buggy?

Reply:
Maybe we should call it a bug, we need to define when we actually want to warn.

Currently, after performing the rewrite, the compiler checks if the class `D` or its constructor is deprecated. In my example, `o.D(1)` is not rewritten to `new o.D(1)` because the constructor is not accessible.

In principle the warning is made up. Given `@deprecated case class C(x: Int)`, neither the companion nor the apply method are deprecated, which lead to the complaint in https://github.com/scala/bug/issues/9650.

We could
  1. copy the annotation from the class to the synthetic companion
  1. copy the annotation from the constructor to the apply method
  1. copy the annotation from the class to the apply method - IMO that is wrong
  1. issue a deprecation warning for `C.apply` when `C` is deprecated; basically what we do today, but make it independent of the rewrite

The last option is the least breaking change, but it leaves the launding path open: `@deprecated case class C(x: Int); val CC = C; CC(1)` won't warn. If we want to change that, then IMO 1. (and possibly 2.) is the right solution.

WDYT? I can change this PR to do 4. and leave 1. / 2. as future work.

I'm pretty convinced that the current warning for using `Alias.apply` (which this PR removes) is not what we want.

Reply:
I meant to return to this question, but instead had a Scala 3 day. I finally read about the experimental features and then learned enough to submit a simple bug fix. You know what's annoying are untyped trees. It's after midnight, so I'll have to sleep on this for now.

Reply:
So you think we should do 1. and 2. form my list above?

I'm on board if possible, we need to get a snese how much pain this will cause. Also it needs to be coordinated with Scala 3, I'll bring it up in the ticket / meeting.

Reply:
yes, also unapply bzw unapplySeq